### PR TITLE
TRITON_ATTN support for KV cache dtype fp8 on sm75 to pre-sm89

### PR DESCRIPTION
**Pull request by mikekg**
# TRITON_ATTN support for KV cache dtype fp8 on sm75 to pre-sm89

## Purpose

Enable `--kv-cache-dtype fp8` on Turing (SM75) and Ampere (SM80/86) with `--attention-backend TRITON_ATTN`, without any change to triton-lang.

SM75–SM88 have no native `fp8e4nv` cast; Triton kernels that emit `.to(tl.float8e4nv)` fail at compile time with *"type fp8e4nv not supported in this architecture"*. This PR adds direct `fp8e4m3 ↔ {fp16, bf16}` conversion via `tl.inline_asm_elementwise` so the KV cache can be stored and loaded as fp8 on Turing/Ampere. The conversion targets the activation dtype: direct `fp8 ↔ fp16` on SM75+ (Turing has no bf16 hardware type) and direct `fp8 ↔ bf16` on SM80+. On SM89+ the native `fp8e4nv` cast is used and this path is bypassed.

The conversion is exposed through one unified pair of entry points, with a single shipped algorithm — round-to-nearest-even encode, exact prmt-LUT decode:

- `vllm/v1/attention/ops/fp8e4nv.py` (new): `convert_to_fp8e4m3(x)` (encode — dispatches on `x.dtype` and **asserts** a 16-bit input, never silently casting) and `convert_from_fp8e4m3(x, dtype)` (decode to the requested dtype). The dtype branch is resolved at Triton compile time.
- `vllm/v1/attention/ops/fp8e4nv_fp16.py`, `fp8e4nv_bf16.py`: the `pack=4` SM75-legal integer-PTX literals (`and/or/shl/shr/add/sub/min/max/setp/selp/prmt/cvt.u32.u16` — no native fp8 cvt, no bf16, no video-SIMD) backing the RNE encode and the exact decode for each dtype. fp16 serves SM75 (no bf16 hardware type); bf16 serves SM80–88.
- `triton_attn.py`: gates `_fp8_software_conv` (SM75–88, CUDA); warns rather than raises; keeps the query unquantized on the software path (a query→fp8e4nv cast would fuse into the RoPE inductor kernel and fail to compile pre-SM89).
- `triton_reshape_and_cache_flash.py`: store path casts the scaled tile to the activation dtype and encodes via `convert_to_fp8e4m3` into the uint8 cache.  (The scaled tile `key / k_scale` is fp32 — a 16-bit value over an fp32 scale promotes — so casting to the activation dtype first selects the correct per-arch encoder.)
- `triton_unified_attention.py`: `_cast_kv_tile` decodes fp8 to the activation dtype via `convert_from_fp8e4m3` and applies the per-tensor scale in that dtype.

## Test Plan

Serve sanity (activation ∈ {bf16, fp16} × KV ∈ {bfloat16, float16, fp8}), and end-to-end throughput + perplexity via the precision sweep, on **A100 (SM80), H100 (SM90), B200 (SM100), and T4 (SM75)**, `--attention-backend TRITON_ATTN`, NVFP4 weights, batched `ppl + 1024+1024 + 8192+8192` per job. **TRITON_ATTN is force-selected for every run** (`--attention-backend TRITON_ATTN`, not the auto-chosen backend): this PR's fp8 software conversion lives in the Triton attention kernels, so the experiment deliberately characterizes the impact of fp8 KV-cache compression on these workloads **specifically under Triton attention**. Each KV dtype is **requested explicitly** and the resolved `kv_cache_dtype` is confirmed from each run's engine-config log — `--kv-cache-dtype auto` is **not** used as the 16-bit baseline, because for these NVFP4 models `auto` resolves to `fp8` (so an `auto`-vs-`fp8` comparison is fp8-vs-fp8).

### 16-bit-native baseline (bfloat16 on SM80+, float16 on T4) vs fp8:
```
python benchmarks/benchmark_throughput.py \
  --model nvidia/Gemma-4-26B-A4B-NVFP4 --attention-backend TRITON_ATTN \
  --kv-cache-dtype bfloat16 --input-len 8192 --output-len 8192 --num-prompts 10   # baseline
python benchmarks/benchmark_throughput.py \
  --model nvidia/Gemma-4-26B-A4B-NVFP4 --attention-backend TRITON_ATTN \
  --kv-cache-dtype fp8 --input-len 8192 --output-len 8192 --num-prompts 10   # fp8
```

## Test Result

The unified conversion and the attention/store integration are validated on A100 (SM80) for both activation dtypes:

- **Conversion suite** (`tests/kernels/attention/test_fp8e4nv_conversion.py`): `4 passed` — decode exact over all 254 finite fp8 bytes and RNE encode bit-exact vs the saturating `torch.float8_e4m3fn` reference, for fp16 and
bf16; the SM89+ native cross-checks skip on SM80.
- **Flash-attention integration** (the merged `convert_to/from` inside the live `unified_attention` + `reshape_and_cache` kernels): the correctness gate passes — fp8-software attention vs a torch reference is `max_abs_err = 0.0071` (native bf16 `0.0010`); `unified_attention` native(bf16) 303 µs vs fp8-software 426 µs; `reshape_and_cache` store native 123 µs vs fp8 ~135 µs (round-mode-independent now that encode is RNE-only).

## Design Exploration and Design Tradeoff Analysis

The single shipped algorithm above was chosen after characterizing the alternatives — three encode rounding modes (`rn`/`rne`/`trunc`) and three decode implementations (arithmetic `region`, byte-LUT `lut`, low-register prmt-LUT `lut2`) — at three levels (standalone microbench, the flash-attention kernel, and full end-to-end). Those commits remain in this branch's history.

### Conversion correctness

`rne` is **bit-exact** vs `torch.float8_e4m3fn` over all finite inputs (all 254 finite fp8 bytes for decode; all 65,536 fp16/bf16 patterns for encode, ±inf/±NaN saturating); `rn` is within 1 fp8-code ULP and `trunc` within 2; decode is exact. On SM89+ the same checks run against the native fp8 cvt.

### Conversion throughput (standalone kernel, pack=4, GB/s, single-stream / 4-stream)

| | rn | rne | trunc |
|---|---|---|---|
| A100 (SM80) fp16 | 540 / 794 | 465 / 643 | 612 / 960 |
| A100 (SM80) bf16 | 541 / 795 | 466 / 644 | 569 / 858 |
| T4 (SM75) fp16 | 127 / 139 | 191 / 213 | 210 / 242 |

Standalone, all decode variants run ~930–957 GB/s on A100 with no meaningful spread (the bare kernel has spare registers). The low-register decode (`lut2`, 35 reg vs the naïve LUT's 46) only pays off **inside** the attention kernel, where register pressure exists.

### Flash-attention kernel — conversion integrated (decode-heavy: 64 seqs, kv_len 2048, GQA 8/2, head 128)

| kernel | A100 | T4 |
|---|---|---|
| `unified_attention`, native KV | 301.8 µs | 4069 µs |
| `unified_attention`, fp8 KV (`lut2` decode) | 483.9 µs (**1.60×**) | 3812 µs(**0.94×**) |
|  `reshape_and_cache`, native store | 118.2 µs | 582 µs |
| `reshape_and_cache`, fp8 store `rne`/`rn`/`trunc` | 132.7 / 132.8 / 131.2 µs | 581 / 574 / 580 µs |
| in-attention decode `lut2` vs `lut` | no difference (128-reg ceiling) | 4097 vs 4172 µs (~1.8%) |

The fp8 store is rounding-mode-neutral on both archs. The compute-rich A100 pays ~60% on the attention kernel for the software decode; the bandwidth-bound T4 *gains* (0.94×) because halving KV traffic outweighs the decode.

### End-to-end — fp8 KV vs 16-bit-float native (bfloat16 SM80+, float16 T4), TRITON_ATTN

Single-request **peak-latency** out_tok/s (`SEQUENTIAL` schedule, num-prompts=10), NVFP4 weights, TP1 (super TP2). Each dtype requested explicitly; resolved `kv_cache_dtype` confirmed per run. One run per cell.

| model | A100/SM80 16b → fp8 | H100/SM90 16b → fp8 | B200/SM100 16b → fp8 | T4/SM75 16b → fp8 |
|---|---|---|---|---|
| nano9b (hybrid-Mamba, 4 attn layers) | 152.5 → 148.9 (−2%) | — | — | ~24.4 → ~24.5 (≈) |
| nano30 | 218.4 → 211.4 (−3%) | — | — | n/a (19 GB > 16) |
| super (TP2) | 93.4 → 91.2 (−2%) | — | — | n/a |
| gemma26b (MoE) | 113.3 → 87.6 (**−23%**) | 164.4 → 169.7 (≈) | 209.7 → 207.8 (≈) | n/a (18 GB > 16) |
| gemma31b (dense) | 36.2 → 30.9 (−15%) | 55.6 → 55.6 (≈) | 79.5 → 78.9 (≈) | n/a |
| diffgemma26b (block-diffusion) | 107.0 → 57.5 (**−46%**) | 247.2 → 200.0 (−19%) | 371.4 → 318.3 (−14%) | n/a |

**Reading the table.** Run-to-run noise on this metric is ~1–2% (one cell, gemma26b/H100, even shows fp8 *faster* than bf16 — unphysical, i.e. noise). So treat every **"≈"** and the **nano/super −2–3%** as *no measurable fp8 cost*; only the deltas ≳10% (A100 gemma/diffgemma; diffgemma on H100/B200) are confidently real. "—" = not run this sweep; "n/a" = does not fit 16 GB. nano/super were run on A100 only; gemma on A100/H100/B200; T4 fits only nano9b.

**Accuracy (perplexity, wikitext-2, 16b / fp8):** nano9b 15.257 / 15.268 · nano30 13.371 / 13.410 · super 5.488 / 5.483 — fp8 is within **≤0.3%** of 16-bit. The three encode rounding modes (`rne`/`rn`/`trunc`) are e2e-equivalent (±0.3% ppl, identical throughput) — `rne` (bit-exact) is the safe default. (The gemma wikitext-ppl path returns garbage for **both** bf16 and fp8 in this harness, so gemma ppl is omitted — a harness artifact, not an fp8 effect.)

### Analysis

There are two separate questions: **which conversion algorithm**, and **whether to convert at all**. The data answers them very differently.

#### (1) The conversion-algorithm choice: encode is e2e-free, the decode is not 

To decide the algorithm on evidence rather than microbench, all six `{rne, rn, trunc} × {lut2, region}` combinations were run **end-to-end** on the three attention-heavy NVFP4 models — the regime where the fp8-KV conversion is most expensive (gemma 15–46% on the cross-arch table above), so any per-algorithm difference is at its largest. A100 (SM80, the software-conversion path), `--attention-backend TRITON_ATTN`, `--kv-cache-dtype fp8`, NVFP4 weights, TP1, 8192+8192, single-request out_tok/s (one run/cell; the engine vLLM build SHA and the per-cell `encode rounding:` / `decode impl:` selection were confirmed from each run's log):

The two axes are independent. The three **encode** modes trade accuracy for integer-op count: `rne` (round-to-nearest-even) is bit-exact vs `torch.float8_e4m3fn` but spends extra ops resolving the even-tie case, `rn` (round-to-nearest, ties-away) is within 1 fp8-code ULP, and `trunc` (round-toward-zero) simply drops the residual mantissa bits for the lowest op count at ≤2 ULP. The two **decode** implementations differ in *how* they reconstruct the 16-bit value from an fp8 byte: `region` does it **arithmetically** — branching on the exponent into the normal / subnormal / overflow ranges and rebuilding the bits with shifts, adds, and predicated= `selp` selects — which is correct but ALU- and register-heavy; `lut2` instead treats the conversion as a **byte lookup**, using the SM75-legal `prmt` byte-permute as the lookup table for the result bytes (exact including subnormals) and folding the byte-isolate-and-shift into the `prmt` selector, so the whole decode costs **35 registers vs the naïve byte-LUT's 46**. Register pressure is the deciding axis: the decode runs *inside* the attention kernel, whose live Q/K/V tiles and softmax accumulators already crowd the register file, so the `region` decode's extra live values throttle occupancy on attention-heavy work — which is precisely why the lower-register `lut2` pulls ahead end-to-end below, even though the standalone microbench (spare registers, no pressure) showed the decodes as equal.

| model | `lut2` rne / rn / trunc | `region` rne / rn / trunc |
|---|---|---|
| gemma26b (MoE) | 86.5 / 88.3 / 86.6 | 82.9 / 82.9 / 84.0 |
| gemma31b (dense) | 30.8 / 30.5 / 30.6 | 29.6 / 29.6 / 29.7 |
| diffgemma26b (block-diffusion) | 59.6 / 59.6 / 61.2 | 53.2 / 53.9 / 53.2 |

**Encode rounding is e2e-free.** With the decode held fixed, `rne` / `rn` / `trunc` agree to ≤~1–2% on every model — run-to-run noise, with no systematic ordering (within-run variance flags the one apparent outlier, diffgemma
`trunc·lut2` +2.7%, as a high-variance cell, not a real gain). This is consistent with the rounding-mode-neutral store kernel (131–133 µs, <1% spread) and the ±0.3% ppl above. So encode rounding carries no e2e penalty — **ship `rne`** (bit-exact, no overhead relative to lower precision choices).

**The decode does *not* come free — `lut2` is materially faster than `region`.** Holding the encode fixed, the shipped low-register prmt-LUT decode (`lut2`) beats the region-arithmetic decode (`region`) end-to-end by **~4% on gemma dense/MoE and ~10% on diffgemma**. The standalone microbench hid this (all decodes ~930–957 GB/s with spare registers; even the in-kernel `lut2`-vs-`lut` comparison showed no A100 difference at the 128-register
ceiling) — but the relevant comparison is `lut2` vs the *arithmetic* `region` decode on attention-heavy models, and there the region decode's higher op/register cost lands on the critical path. diffgemma makes it largest: its
block-diffusion schedule re-reads the KV many times per token (see §(2)), so each extra decode instruction is paid repeatedly, widening the gap to ~10%. This is the measured justification for the shipped decode: **`lut2` is not merely "lower-register at no penalty" — it is 4–10% faster than the alternative exactly where attention dominates.**

So the algorithm choice is settled by data, not preference: **ship `rne`\ (free, bit-exact) + `lut2` (4–10% faster than `region` on attention-heavy decode).**

#### (2) The conversion *cost* (fp8-decode vs 16-bit no-decode) = per-read decode expense × read count

`_cast_kv_tile` widens fp8 KV back to the activation dtype on *every* KV-tile read (the dot runs in bf16/fp16, not fp8). The cost is that decode, paid once per read. Three regimes:

- **SM75–88 — this PR's software conversion path (the path under review).** Standalone the software decode adds ~60% to the attention kernel on A100 (`unified_attention` 301.8 → 483.9 µs); end-to-end that dilutes to a model-level cost **scaling with attention weight** — gemma dense/MoE 15–46%, while hybrid-Mamba nemotron-nano (4 attention layers → KV touches little compute) is **≤2–3% and not worth analyzing further**. On the bandwidth-bound **T4 the software decode is a net win** — fp8-KV attention is *faster* than fp16 (0.94×), since halving KV traffic outweighs the decode.

- **SM89+ (H100/B200) — native cvt; shown as a control, pre-existing and not this PR's path.** Here fp8 KV uses the hardware cvt, so these numbers reflect the conversion-*path* overhead with the **native cvt** — a single fp8↔bf16 cvt instruction per KV read: quasi-free, but still imposing that instruction's cost plus whatever schedule disruption it causes in the kernel. Its per-read cost is small enough to sit within run-to-run noise for autoregressive decode (gemma26b/gemma31b fp8 ≈ bf16), but it is real and accumulates: diffgemma's −14–19% below is literally that native-cvt cost paid over the block-denoising re-reads. This is the clean control proving the A100 cost is specifically the *software decode*: same model, same path, only the decode expense changes (diffgemma −46% A100 software vs −14–19% H100/B200 native cvt).

- **diffgemma — the diffusion outlier.** diffgemma generates by **block-diffusion denoising rather than autoregressive decode**: it re-attends over the block's KV at every denoising step, so each KV element is decoded\ *several times per output token* instead of once as in autoregressive generation. That is why it carries a conversion cost on **every** arch — including H100/B200, where the native cvt's per-read *overhead* on the autoregressive gemmas is within noise (fp8 ≈ bf16): even that single-instruction cvt, multiplied by the block-denoising re-read count, accumulates to −14–19% (and to −46% under the A100 software decode). This is a property of the block-diffusion attention schedule, not of the conversion code; it is the one model where the KV-decode path stays on the critical path regardless of cvt cost.

#### Net

fp8 KV delivers its intended win — ~2× KV-cache capacity (1 byte/element vs 2) at ≈-or-better throughput — exactly where it should: bandwidth-bound **T4** (about same throughput -- marginally but repeatably faster), **SM89+** (the fp8 *conversion* is quasi-free → fp8 ≈ bf16 throughput), and low-attention models (**nano ≤2%**). The throughput cost is material only for attention-heavy autoregressive decode on the SM75–88 software path (**gemma on A100**, 15–46%) and for the block-diffusion pattern (**diffgemma**, all archs) — and even there it is often the right tradeoff for a KV cache: the alternative, serving the same 2× context with 16-bit KV, needs ~2× the KV memory and therefore more GPUs, which is almost always a higher expense and is not even possible in every environment. Trading some attention throughput for 2× KV capacity on the *same* hardware is frequently the cheaper option.

AI assistance (Claude Code) in the preparation of this PR.

### CODE REVIEW

File: vllm/model_executor/layers/quantization/base_config.py [hunk 1]
```diff
@@ -48,11 +48,18 @@ def embedding(self, layer: torch.nn.Module, *args, **kwargs) -> torch.Tensor:
         raise NotImplementedError
 
     # Not required functions
-    def tie_weights(self, layer: torch.nn.Module, *args, **kwargs):
```
Comment (mikekg) [inline vllm/model_executor/layers/quantization/base_config.py]: [#45544](https://github.com/vllm-project/vllm/pull/45544) — `[Bugfix] Default tie_weights to sharing the weight for quantized embeddings`

**Comment by mikekg** [comment]
I would appreciate reviewer's feedback on structure -- should we move the conversion ops into a common submodule available to other operations, or keep them here under ops?

File: vllm/v1/attention/ops/triton_unified_attention.py [hunk 2]
```diff
@@ -49,6 +53,12 @@ def _cast_kv_tile(data, Q, tensor_scale, KV_QUANT_MODE: tl.constexpr):
       into the attention score and output accumulator.
     """
     if KV_QUANT_MODE == 1:
+        if FP8_SOFTWARE_CONV:
+            # Software-decode the fp8 bytes to the activation dtype and apply the
+            # per-tensor scale in that dtype (no fp32).
+            if Q.dtype == tl.float16:
```
Comment (mikekg) [inline vllm/v1/attention/ops/triton_unified_attention.py]: or make it `fp8e4m3(data, Q.dtype)` or even `fp_dec(data, Q.dtype` so it might cover e5m2 in the same function and dispatch internally ?
